File: client/response.cpp
/*
Maman 15
Author: Elad Sheffer
*/

#include "response.h"

/// <summary>
/// Ctor
/// </summary>
Response::Response() {
	this->version = 0;
	this->code = 0;
	this->payload = "";
	this->payloadSize = 0;
}

/// <summary>
/// Ctor
/// </summary>
/// <param name="version"></param>
Response::Response(uint8_t version) {
	this->version = version;
	this->code = 0;
	this->payloadSize = 0;
	this->payload = "";
}

Response::Response(uint8_t version, uint16_t code, std::array<unsigned char, UUID_LENGTH_RESPONSE> clientId, std::string payload) {
	this->version = version;
	this->code = code;
	this->payload = payload;
	this->payloadSize = (uint32_t)payload.size();
	this->clientId = clientId;
}

/// <summary>
/// Dtor
/// </summary>
Response::~Response() {
	this->version = 0;
	this->code = 0;
	this->clientId = { 0 };
	this->payload = "";
	this->payloadSize = 0;
}

/// <summary>
/// Returns version
/// </summary>
/// <returns>version</returns>
uint8_t Response::getVersion() {
	return this->version;
}

/// <summary>
/// Returns code
/// </summary>
/// <returns>code</returns>
uint16_t Response::getCode() {
	return this->code;
}

/// <summary>
/// Returns size of payload
/// </summary>
/// <returns>size of payload</returns>
uint32_t Response::getPayloadSize() {
	return this->payloadSize;
}

/// <summary>
/// Returns user id
/// </summary>
/// <returns>user id</returns>
std::array<unsigned char, UUID_LENGTH_RESPONSE> Response::getClientId() {
	return this->clientId;
}

/// <summary>
/// Returns payload
/// </summary>
/// <returns>payload</returns>
std::string Response::getPayload() {
	return this->payload;
}

/// <summary>
/// Sets code
/// </summary>
/// <param name="code"></param>
void Response::setCode(uint16_t code) {
	this->code = code;
}

/// <summary>
/// Sets payload
/// </summary>
/// <param name="payload"></param>
void Response::setPayload(std::string payload) {
	this->payload = payload;
	this->payloadSize = (uint32_t)payload.size();

}

/// <summary>
/// Parses Request Header
/// </summary>
/// <param name="requestHeader"></param>
/// <returns></returns>
ResponseHeader Response::deparseHeader(Response response) {

	ResponseHeader responseHeader = { 0 };
	responseHeader.responseData.version = response.getVersion();
	responseHeader.responseData.code = response.getCode();
	responseHeader.responseData.payloadSize = response.getPayloadSize();

	return responseHeader;
}

File: server/response.py
# response.py
# Author: Elad Sheffer

import struct

UUID_LENGTH = 16


class Response:
    __version = None
    __code = None
    __client_id = None
    __payload = None
    __payload_size = None

    def __init__(self, version=0, code=0, client_id=b'', payload=b''):
        self.__version = struct.unpack('=B', struct.pack('=B', version))
        self.__code = struct.unpack('=H', struct.pack('=H', code))
        self.__client_id = struct.unpack('=' + str(UUID_LENGTH) + 's', struct.pack('=' + str(UUID_LENGTH) + 's',
                                                                                   client_id))
        self.__payload = payload
        self.__payload_size = struct.unpack('=I', struct.pack('=I', self.__payload.__len__()))

    def get_version(self):
        return self.__version[0]

    def get_code(self):
        return self.__code[0]

    def get_client_id(self):
        return self.__client_id[0]

    def get_payload_size(self):
        return self.__payload_size[0]

    def get_payload(self):
        return self.__payload

    def set_version(self, version):
        self.__version = struct.unpack('=B', struct.pack('=B', version))

    def set_code(self, code):
        self.__code = struct.unpack('=H', struct.pack('=H', code))

    def set_client_id(self, client_id):
        self.__client_id = struct.unpack('=' + str(UUID_LENGTH) + 's', struct.pack('=' + str(UUID_LENGTH) + 's',
                                                                                   client_id))

    def set_payload(self, payload):
        self.__payload = payload
        self.__payload_size = struct.unpack('=I', struct.pack('=I', self.__payload.__len__()))
